### PR TITLE
Corrected warning for homonymous, but not shadowing declarations

### LINKED ISSUE #4558: Wrong warning on shadowed declaration
Solidity prints the declaration shadowed warning on a contract like 
```solidity
contract A {
  function foo() external { }
  function bar() external {
    uint foo = 10; // warning that internal variable `foo` shadows the function `foo`.
  }
}
```

But actually, because function `foo` is external, it is not internally accessible in `bar` function unless `this.foo()` or `a.foo()` is used, so I think function `foo` is not shadowed by internal variable `foo`.

### PR DESCRIPTION
**Pull request by a3d4**
Fixes #4558.

### CODE REVIEW

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 1]
```diff
@@ -503,11 +503,20 @@ bool DeclarationRegistrationHelper::registerDeclaration(
...
-				"This declaration shadows an existing declaration.",
-				SecondarySourceLocation().append("The shadowed declaration is here:", shadowedLocation)
+
+			if (shadowedDeclaration->visibility() == Visibility::External)
+				_errorReporter.warning(
+					8760_error,
+					_declaration.location(),
+					"This declaration has the same name as an existing external declaration.",
+					SecondarySourceLocation().append("The external declaration is here:", shadowedLocation)
+				);
```
Comment (a3d4) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: Is there a legitimate case when a declaration with `External` visibility is really shadowed by another declaration?

**Comment by chriseth** [comment]
Can you check this, please?
```
function e() {}
contract test {
    function f() public pure { uint e; uint t; }
}
funciton t() {}
```

**Comment by a3d4** [comment]
```
function e() {}
contract test {
    function f() pure public { uint e; uint g; uint h; }
    function g() pure public {}
}
function h() {}
```

We warn about shadowiing `e()` only. `g()` and `h()` go unnoticed.

I also saw that we do not seem to consider declarations which go after the current one. I wanted to address it as a separate issue when this PR is completed.

**Comment by chriseth** [comment]
Right, the "no shadowing warning for later declarations" is a different issue.

**Comment by chriseth** [comment]
Please add a changelog entry - fine otherwise!

**Review by chriseth — COMMENTED** [review]
Actually wait a second:

Do you think a solution including the "invisible" flag would be possible? The "invisible" flag is meant as a way to say that the identifier is present but not visible - which is basically exactly what we want.

**Comment by chriseth** [comment]
Oh and maybe `_declaration.isVisibleInContract()` would also be better than checking for external visibility - which is a quite specific property.

**Comment by a3d4** [comment]
> Oh and maybe _declaration.isVisibleInContract() would also be better

Done.

> Do you think a solution including the "invisible" flag would be possible?

Do you mean `bool _invisible` in `DeclarationContainer::registerDeclaration` and related things? It could make sense, let me have a closer look.

**Comment by a3d4** [comment]
@chriseth 

> Do you think a solution including the "invisible" flag would be possible?

Consider
```
contract C {
    function f() public pure {
        uint x;
        { uint x; }
    }
}
```

If `invisible = (!_declaration.isVisibleInContract() || _inactive)`, then both `x` are invisible, right (because they are in the block scope, hence inactive)? Then the warning would be "This declaration has the same name as another declaration.", while it should be "This declaration shadows an existing declaration.".

Or did I misunderstand you?

**Comment by chriseth** [comment]
I think the error looks fine. It's maybe not that important after all. Also in your example - https://github.com/ethereum/solidity/pull/9703/files#diff-5d15f07f01f7bb76a8c6410ee3b1a6e1R6 - I would say this should be shadowing, since they are different scopes.

But to sum up: I think it is better to use a flag native to the NameAndTypeResolver instead of something that is only usable inside contracts.

**Comment by a3d4** [comment]
> I think the error looks fine. It's maybe not that important after all. Also in your example - https://github.com/ethereum/solidity/pull/9703/files#diff-5d15f07f01f7bb76a8c6410ee3b1a6e1R6 - I would say this should be shadowing, since they are different scopes.

I'm a bit confused. Do you mean **name_shadowing2.sol** or **name_pseudo_shadowing.sol**?

**name_shadowing2.sol** already reports "shadowing", and **name_pseudo_shadowing.sol** is _not_ shadowing, according to #4558.

What do we mean by "shadowing" exactly? My rough definition would be that "new x" shadows "old x" iff removing "new x" would not break compilation, and all references of "new x" would become references to "old x".

**Review by chriseth — APPROVED** [review]
I just noticed that we use `_declaration.isVisibleInContract` already in that function, so this is fine! Sorry for the confusion!

**Pull request merged**